### PR TITLE
Fix issue 20267 - Error: `string` is used as a type - and similar “smart” error messages

### PR DESCRIPTION
**Pull request by 0xEAB**
This PR adds a special case to `visitIdentifier()` that produces a better understandable error message in case a variable is used as a type (which can happen accidentally). I couldn't really come up with any other relevant edge case that would trigger a similar confusing error message, so I went for the special case when `e.op == TOK.variable`.

For obvious reasons, this change requires tests that check for this error message to be adapted accordingly. I also added a new specialized test.

### CODE REVIEW

**Comment by 0xEAB** [comment]
>  any other relevant edge case

So why not *just* go for the general approach?

Well, while it would be trivial to implement
```diff
-                .error(loc, "var `%s` is used as a type", mtype.toChars());
+                .error(loc, "%s `%s` is used as a type", Token.toChar(e.op), mtype.toChars());
```

...it would result in confusing error messages either.
https://github.com/0xEAB/dmd/blob/5404698f294ee43f842951593f92cea89e536c60/test/fail_compilation/ice11849b.d#L5
This would end up as:
```
Error: error `DWORD1` is used as a type
```
(*"oh nice to know that an error whatsoever cannot be used as a type - definitely not obvious that an errorr thingy isn't a type... so what did I do wrong now?"* `// sarcasm off`)

To fix this we'd need to replace my whitelist with a blacklist (`!= TOK.error`) which would probably make everything even more complicated.

**Comment by thewilsonator** [comment]
Thanks. I think it would be better to say `Variable` rather than `var`. Also it would be nice to point to the declaration site of the variable in question, something like:
```d
{
    .error(loc, "var `%s` is used as a type", mtype.toChars());
    e.errorSupplimental("Declared here");
}
```

**Comment by 0xEAB** [comment]
> I think it would be better to say `Variable` rather than `var`.

`Token.toChar()` also maps `TOK.variable` to `var`. That's why I went for it, more consistent imho.

**Comment by 0xEAB** [comment]
> Also it would be nice to point to the declaration site of the variable in question

Done. Went for a supplemental message similar to:
https://github.com/0xEAB/dmd/blob/5404698f294ee43f842951593f92cea89e536c60/src/dmd/expression.d#L1256

File: test/fail_compilation/typeerrors.d [hunk 1]
```diff
@@ -2,22 +2,23 @@
...
+fail_compilation/typeerrors.d(47): Error: var `x` is used as a type
+fail_compilation/typeerrors.d(38):        var `x` is declared here
+fail_compilation/typeerrors.d(48): Error: cannot have associative array key of `void()`
+fail_compilation/typeerrors.d(49): Error: cannot have associative array key of `void`
+fail_compilation/typeerrors.d(50): Error: cannot have array of scope `typeerrors.C`
+fail_compilation/typeerrors.d(51): Error: cannot have associative array of `void`
+fail_compilation/typeerrors.d(52): Error: cannot have associative array of `void()`
+fail_compilation/typeerrors.d(54): Error: cannot have parameter of type `void`
+fail_compilation/typeerrors.d(56): Error: slice `[1..5]` is out of range of [0..4]
+fail_compilation/typeerrors.d(57): Error: slice `[2..1]` is out of range of [0..4]
```
Comment (0xEAB) [inline test/fail_compilation/typeerrors.d]: looks bigger than it is, just had to update all those line numbers

**Comment by thewilsonator** [comment]
> Token.toChar() also maps TOK.variable to "var". 

I'd still prefer variable, it doesn't matter what name the compiler uses internally, but what appears in other error messages, e.g. I'm fairly sure that "function" appears a lot, but "func" doesn't.

**Comment by 0xEAB** [comment]
> e.g. I'm fairly sure that "function" appears a lot, but "func" doesn't.

yeah but that's because of:
```d
TOK.function_: "function",
```

**Comment by 0xEAB** [comment]
I mean, don't get me wrong. I'm actually fine with both options. (Don't take my reasoning for the option I initially went with, as what I'd prefer, please.)

Maybe we should also address those (potential) inconsistency issues in the future.

**Comment by WebFreak001** [comment]
Suggestion: there should be a test which imports some variable from some other package (like `import std.system;` which defines a global variable called "endian")

If a test using `endian foo = 1;` only shows the definition in std.system, it would be very cool to have an additional line saying "variable imported here" on the import line

**Comment by 0xEAB** [comment]
```d
fail_compilation/test20267.d(27): Error: var `endian` is used as a type
/home/me/some/fancy/path/dlang/dmd/test/../../phobos/std/system.d(74):        variable `endian` is declared here
```
I'm afraid I don't know how to add this to the testsuite in a reasonable way.

**Comment by 0xEAB** [comment]
We'll need an extra edge-case for this. Like @WebFreak001 said:
> yeah it would be really cool if it showed "imported from here"

**Comment by 0xEAB** [comment]
Which error message would you prefer?

- ```d
  fail_compilation/test20267.d(27): Error: var `array` is used as a type
  fail_compilation/test20267.d(27):        variable `array` is imported from there: imports.test20267
  ```

- ```d
  fail_compilation/test20267.d(27): Error: var `array` is used as a type
  fail_compilation/imports/test20267.d(3):        variable `array` is imported from here: imports.test20267
  ```

- ```d
  fail_compilation/test20267.d(27): Error: var `array` is used as a type
  fail_compilation/imports/test20267.d(3):        variable `array` is imported from here
  ```

**Comment by 0xEAB** [comment]
@WebFreak001 gave me a better suggestion:

> can't it have multiple error message lines?
> so it shows:
> ```
> variable `array` is used as a type
>     variable `array` is imported from here
>     variable `array` is defined here
> ```

File: test/fail_compilation/test20267.d [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+PERMUTE_ARGS:
+TEST_OUTPUT:
+---
+fail_compilation/test20267.d(20): Error: variable `string` is used as a type
+fail_compilation/test20267.d(19):        variable `string` is declared here
+fail_compilation/test20267.d(23): Error: variable `boolean` is used as a type
+fail_compilation/test20267.d(22):        variable `boolean` is declared here
+fail_compilation/test20267.d(30): Error: variable `array` is used as a type
+fail_compilation/test20267.d(28):        variable `array` is imported here from: `imports.test20267`
+fail_compilation/imports/test20267.d(3):        variable `array` is declared here
```
Comment (0xEAB) [inline test/fail_compilation/test20267.d]: Here we go :grin:

**Comment by thewilsonator** [comment]
Looks nice, is this good to go?

**Comment by 0xEAB** [comment]
As far as I'm concerned, it's ready.
